### PR TITLE
[Merged by Bors] - feat: `LinearEquiv.congrQuadraticMap` and `LinearEquiv.congrRight₂`

### PR DESCRIPTION
**Pull request by mans0954**
`LinearEquiv.congrQuadraticMap` matches `LinearMap.compQuadraticMap`, while `LinearEquiv.congrRight₂` somewhat matches `LinearMap.compr₂`.

Used in  #14988

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 1]
```diff
@@ -579,6 +579,30 @@ def _root_.LinearMap.compQuadraticMap' [CommSemiring S] [Algebra S R] [Module S
...
+  right_inv _ := ext fun x => by
+    simp only [LinearMap.compQuadraticMap_apply, LinearEquiv.coe_coe, LinearEquiv.apply_symm_apply]
+  map_add' _ _ := ext fun x => by
+    simp only [LinearMap.compQuadraticMap_apply, add_apply, map_add, LinearEquiv.coe_coe]
+  map_smul' _ _ := ext fun x => by
+    simp only [LinearMap.compQuadraticMap_apply, smul_apply, LinearMapClass.map_smul,
+      LinearEquiv.coe_coe, RingHom.id_apply]
+
+@[simp]
+theorem congr₂_refl :
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:600]: Could you add these for the bilinear version too?

File: Mathlib/LinearAlgebra/BilinearForm/Hom.lean [hunk 2]
```diff
@@ -305,6 +306,15 @@ theorem comp_congr (e : M' ≃ₗ[R] M'') (B : BilinForm R M) (l r : M' →ₗ[R
       B.comp (l.comp (e.symm : M'' →ₗ[R] M')) (r.comp (e.symm : M'' →ₗ[R] M')) :=
   rfl
 
+variable {N₁ N₂ : Type*}
+variable [AddCommMonoid N₁] [AddCommMonoid N₂] [Module R N₁] [Module R N₂]
+
+/-- When `N₁` and `N₂` are equivalent, bilinear maps on `M` into `N₁` are equivalent to bilinear
+maps into `N₂`. -/
+@[simps!]
+def _root_.LinearMap.BilinMap.congr₂ (e : N₁ ≃ₗ[R] N₂) : BilinMap R M N₁ ≃ₗ[R] BilinMap R M N₂ :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/BilinearForm/Hom.lean:315]: Do we have a the non-equiv spelling of this? If so, a lemma linking then would be good.

(In the quadratic case `simps` I believe generates such a lemma, but I will check more closely when I'm next at lean)

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean (hunk 1 above)
Comment (mans0954) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:600]: Done, thanks.

File: Mathlib/LinearAlgebra/BilinearForm/Hom.lean (hunk 2 above)
Comment (mans0954) [reply Mathlib/LinearAlgebra/BilinearForm/Hom.lean]: `@[simps!?]` outputs:
```
[simps.verbose] adding projection LinearMap.BilinForm.LinearMap.BilinMap.congr₂_symm_apply_apply_apply:
      ∀ {R : Type u_1} {M : Type u_2} [inst : CommSemiring R] [inst_1 : AddCommMonoid M] [inst_2 : Module R M]
        {N₁ : Type u_9} {N₂ : Type u_10} [inst_3 : AddCommMonoid N₁] [inst_4 : AddCommMonoid N₂] [inst_5 : Module R N₁]
        [inst_6 : Module R N₂] (e : N₁ ≃ₗ[R] N₂) (f : M →ₗ[R] M →ₗ[R] N₂) (a a_1 : M),
        (((LinearMap.BilinForm.LinearMap.BilinMap.congr₂ e).symm f) a) a_1 = e.symm ((f a) a_1) 

[simps.verbose] adding projection LinearMap.BilinForm.LinearMap.BilinMap.congr₂_apply_apply_apply:
      ∀ {R : Type u_1} {M : Type u_2} [inst : CommSemiring R] [inst_1 : AddCommMonoid M] [inst_2 : Module R M]
        {N₁ : Type u_9} {N₂ : Type u_10} [inst_3 : AddCommMonoid N₁] [inst_4 : AddCommMonoid N₂] [inst_5 : Module R N₁]
        [inst_6 : Module R N₂] (e : N₁ ≃ₗ[R] N₂) (f : M →ₗ[R] M →ₗ[R] N₁) (a a_1 : M),
        (((LinearMap.BilinForm.LinearMap.BilinMap.congr₂ e) f) a) a_1 = e ((f a) a_1) 
```

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/BilinearForm/Hom.lean]: Right, that's what I was afraid of. I added the missing lemma.

**Comment by eric-wieser** [comment]
I'm not super happy with the naming, but I also don't have any better ideas.

Maybe another reviewer does!

maintainer merge

**Comment by kbuzzard** [comment]
If we're going to bikeshed over names then another possibility for `LinearMap.BilinForm.congr₂` would I guess be `LinearMap.BilinForm.congrRight`. But I'd be happy to merge either choice.

**Comment by eric-wieser** [comment]
I think that would be rather confusing, because [`LinearMap.BilinForm.compRight`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/BilinearForm/Hom.html#LinearMap.BilinForm.compRight) sounds like it ought to be analogous, but is about the (right) domain rather than the codomain.

**Comment by jcommelin** [comment]
@mans0954 Unrelated, but github gives the following warning:

> The head ref may contain hidden characters: "mans0954/congr\u2082"

I think it's good to stick to ascii for branch names.

**Comment by mans0954** [comment]
So, is the consensus for `LinearMap.BilinForm.congrRight`? I've pushed a commit with that.

**Comment by mans0954** [comment]
> you get the definition in this PR by swapping `LinearMap` with `BilinMap`, not by swapping `LinearEquiv` with` BilinMap`.

I'm afraid this has been going on for so long I have no recollection of where I adapted it from now.

**Comment by mans0954** [comment]
@eric-wieser is this what you mean?

https://github.com/leanprover-community/mathlib4/pull/17454/commits/ecc76192283f364be42a44e2885910ff8b26a7d4

**Comment by eric-wieser** [comment]
Yes, that's what I meant. @jcommelin, are you happy with those?

bors d=@jcommelin

**Comment by mans0954** [comment]
@jcommelin are you able to merge this PR now please?

**Comment by jcommelin** [comment]
bors merge

**Pull request closed without merging**